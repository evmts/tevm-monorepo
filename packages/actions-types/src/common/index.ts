export type { Abi } from './Abi.js'
export type { Block } from './Block.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockTag } from './BlockTag.js'
export type { Address } from './Address.js'
export type { Hex } from './Hex.js'
export type { EmptyParams } from './EmptyParams.js'
export type { Log } from './Log.js'
export type { BlockResult } from './BlockResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionResult } from './TransactionResult.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { NetworkConfig } from './NetworkConfig.js'
