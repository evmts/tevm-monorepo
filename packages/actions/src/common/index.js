export { isValidCallEvents, validateCallEvents } from './zCallEvents.js'
