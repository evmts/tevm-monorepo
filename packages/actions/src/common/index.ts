export type { Abi } from './Abi.js'
export type { AccountState } from './AccountState.js'
export type { Address } from './Address.js'
export type { Block } from './Block.js'
export type { BlockOverrideSet } from './BlockOverrideSet.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockResult } from './BlockResult.js'
export type { BlockTag } from './BlockTag.js'
export type { CallEvents, Message, NewContractEvent } from './CallEvents.js'
export type { CallTraceResult } from './CallTraceResult.js'
export type { EmptyParams } from './EmptyParams.js'
export type { FlatCallTraceResult, FlatCallTraceCall } from './FlatCallTraceResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { Hex } from './Hex.js'
export type { Log } from './Log.js'
export type { NetworkConfig } from './NetworkConfig.js'
export type { PrestateTraceResult } from './PrestateTraceResult.js'
export type { StateOverrideSet } from './StateOverrideSet.js'
export type { StructLog } from './StructLog.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TransactionResult } from './TransactionResult.js'
