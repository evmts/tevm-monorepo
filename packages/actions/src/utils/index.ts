export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
