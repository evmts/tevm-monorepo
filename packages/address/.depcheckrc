ignores: [
  "ms"
]
skip-missing: true
