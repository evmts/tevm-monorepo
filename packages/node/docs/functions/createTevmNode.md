[**@tevm/node**](../README.md) • **Docs**

***

[@tevm/node](../globals.md) / createTevmNode

# Function: createTevmNode()

> **createTevmNode**(`options`?): [`TevmNode`](../type-aliases/TevmNode.md)\<`"fork"` \| `"normal"`, `object`\>

## Parameters

• **options?**: [`TevmNodeOptions`](../type-aliases/TevmNodeOptions.md)\<`Common`\> = `{}`

## Returns

[`TevmNode`](../type-aliases/TevmNode.md)\<`"fork"` \| `"normal"`, `object`\>

## Defined in

[packages/node/src/createTevmNode.js:27](https://github.com/evmts/tevm-monorepo/blob/main/packages/node/src/createTevmNode.js#L27)
