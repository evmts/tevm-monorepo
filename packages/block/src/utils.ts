import { EthjsAddress, keccak256, toType as originalToType, toBytes } from '@tevm/utils'

/**
 * Creates a zero-filled buffer of the specified length
 * @param {number} length - The length of the buffer
 * @returns {Uint8Array} A zero-filled Uint8Array
 */
export function zeros(length: number): Uint8Array {
	return new Uint8Array(length)
}

/**
 * Creates the zero address
 * @returns {EthjsAddress} The zero address
 */
export function createZeroAddress() {
	return new EthjsAddress(new Uint8Array(20))
}

/**
 * Creates an address from a public key
 * @param {Uint8Array} publicKey - The public key
 * @returns {EthjsAddress} The address
 */
export function createAddressFromPublicKey(publicKey: Uint8Array): EthjsAddress {
	// Create address from public key by taking the last 20 bytes of keccak256 hash
	const addressBytes = keccak256(publicKey, 'bytes').slice(-20)
	return new EthjsAddress(addressBytes)
}

/**
 * Wrapper for toType that handles string inputs properly
 * @param {any} input - The input value
 * @param {number} outputType - The output type
 * @returns {any} The converted value
 */
export function safeToType(input: any, outputType: number): any {
	if (input === null || input === undefined) {
		return undefined
	}

	// Handle string inputs by converting to bytes first
	if (typeof input === 'string' && outputType === 2) {
		const bytes = toBytes(input)
		return originalToType(bytes, outputType as any)
	}

	return originalToType(input, outputType as any)
}

/**
 * Gets the v value from a signature
 * @param {any} signature - The signature object
 * @returns {bigint} The v value
 */
export function getSignatureV(signature: any): bigint {
	// Handle both old format (v) and new format (recovery)
	if (signature.v !== undefined) {
		return BigInt(signature.v)
	}
	if (signature.recovery !== undefined) {
		return BigInt(signature.recovery) + 27n
	}
	throw new Error('Invalid signature format: missing v or recovery')
}
