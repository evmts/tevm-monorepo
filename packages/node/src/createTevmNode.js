import { createChain } from '@tevm/blockchain'
import { createCommon, tevmDefault } from '@tevm/common'
import { createEvm } from '@tevm/evm'
import { createLogger } from '@tevm/logger'
import { ReceiptsManager, createMapDb } from '@tevm/receipt-manager'
import { createStateManager } from '@tevm/state'
import { TxPool } from '@tevm/txpool'
import { KECCAK256_RLP, bytesToHex, hexToBigInt, keccak256 } from '@tevm/utils'
import { createVm } from '@tevm/vm'
import { DEFAULT_CHAIN_ID } from './DEFAULT_CHAIN_ID.js'
import { GENESIS_STATE } from './GENESIS_STATE.js'
import { getBlockNumber } from './getBlockNumber.js'
import { getChainId } from './getChainId.js'
import { statePersister } from './statePersister.js'

// TODO the common code is not very good and should be moved to common package
// it has rotted from a previous implementation where the chainId was not used by vm

/**
 * Creates the base instance of a memory client
 * @param {import('./TevmNodeOptions.js').TevmNodeOptions} [options]
 * @returns {import('./TevmNode.js').TevmNode}
 * @example
 * ```ts
 *  ```
 */
export const createTevmNode = (options = {}) => {
	/**
	 * @type {import('@tevm/utils').Address | undefined}
	 */
	let impersonatedAccount = undefined
	/**
	 * @param {import('@tevm/utils').Address | undefined} address
	 * returns {void}
	 */
	const setImpersonatedAccount = (address) => {
		impersonatedAccount = address
	}

	const loggingLevel = options.loggingLevel ?? 'warn'
	const logger = createLogger({
		name: 'TevmClient',
		level: loggingLevel,
	})

	/**
	 * @returns {Promise<import('@tevm/state').StateOptions>}
	 */
	const getStateManagerOpts = async () => {
		if (options.fork?.transport) {
			// if the user passed in latest we must use an explicit block tag
			const blockTag = await blockTagPromise
			return {
				loggingLevel,
				...(options.persister ? { onCommit: statePersister(options.persister, logger) } : {}),
				fork: {
					...options.fork,
					blockTag,
				},
			}
		}
		return {
			loggingLevel,
			...(options.storageCache !== undefined ? { storageCache: options.storageCache } : {}),
			...(options.contractCache !== undefined ? { contractCache: options.contractCache } : {}),
			...(options.accountsCache !== undefined ? { accountsCache: options.accountsCache } : {}),
			...(options.fork?.transport ? options.fork : {}),
			...(options.persister !== undefined ? { onCommit: statePersister(options.persister, logger) } : {}),
		}
	}

	/**
	 * Create the extend function in a generic way
	 * @param {import('./TevmNode.js').TevmNode} client
	 * @returns {import('./TevmNode.js').TevmNode['extend']}
	 */
	const extend = (client) => (extension) => {
		Object.assign(client, extension(client))
		return /** @type {any}*/ (client)
	}

	const chainIdPromise = (async () => {
		if (options?.common) {
			return options?.common.id
		}
		const forkTransport = options?.fork?.transport
		if (forkTransport) {
			const id = await getChainId(forkTransport)
			return id
		}
		return DEFAULT_CHAIN_ID
	})().then((chainId) => {
		logger.debug({ chainId }, 'Creating client with chainId')
		return BigInt(chainId)
	})

	const blockTagPromise = (async () => {
		if (options.fork === undefined) {
			// this is ultimately unused
			return 0n
		}
		// TODO handle other moving block tags like `safe`
		// we need to fetch the latest block number and return that otherwise we may have inconsistencies from block number changing
		if (options.fork.blockTag === undefined || options.fork.blockTag === 'latest') {
			const latestBlockNumber = await getBlockNumber(options.fork.transport)
			logger.debug({ latestBlockNumber }, 'fetched fork block number from provided forkurl')
			return latestBlockNumber
		}
		return options.fork.blockTag
	})()

	const chainCommonPromise = chainIdPromise
		.then((chainId) => {
			if (options.common) {
				return createCommon({
					...options.common,
					id: Number(chainId),
					loggingLevel: options.loggingLevel ?? 'warn',
					hardfork: /** @type {import('@tevm/common').Hardfork}*/ (options.common.ethjsCommon.hardfork()) ?? 'cancun',
					eips: options.common.ethjsCommon.eips(),
					customCrypto: options.common.ethjsCommon.customCrypto,
				})
			}
			return createCommon({
				...tevmDefault,
				id: Number(chainId),
				loggingLevel: options.loggingLevel ?? 'warn',
				hardfork: 'cancun',
				eips: [],
			})
		})
		.then((common) => {
			// ALWAYS Copy common so we don't modify the global instances since it's stateful!
			const copy = common.copy()
			return copy
		})

	const blockchainPromise = Promise.all([chainCommonPromise, blockTagPromise]).then(([common, blockTag]) => {
		return createChain({
			loggingLevel,
			common,
			...(options.fork?.transport !== undefined
				? {
						fork: {
							transport: options.fork.transport,
							blockTag,
						},
					}
				: {}),
		})
	})

	const stateManagerPromise = blockchainPromise
		.then(async (blockchain) => {
			// wait for blockchain to initialize so we have the correct state root
			await blockchain.ready()
			return blockchain
		})
		.then((chain) => chain.getCanonicalHeadBlock())
		.then(async (headBlock) => {
			const stateRootHex = bytesToHex(headBlock.header.stateRoot)
			const restoredState = options.persister?.restoreState()
			restoredState && console.log(restoredState)
			if (restoredState) {
				logger.debug(restoredState, 'Restoring persisted state...')
				logger.warn(
					'State persistence is an experimental feature. It currently does not persist the blockchain state only the EVM state.',
				)
				/**
				 * @type {import('@tevm/state').TevmState}
				 */
				const parsedState = {}
				for (const [k, v] of Object.entries(restoredState)) {
					const { nonce, balance, storageRoot, codeHash } = v
					parsedState[/** @type {import('@tevm/utils').Address}*/ (k)] = {
						...v,
						nonce: hexToBigInt(nonce),
						balance: hexToBigInt(balance),
						storageRoot: storageRoot,
						codeHash: codeHash,
					}
				}
				// We might want to double check we aren't forking and overwriting this somehow
				// TODO we should be storing blockchain state too
				const manager = createStateManager({
					genesisState: parsedState,
					currentStateRoot: bytesToHex(headBlock.header.stateRoot),
					stateRoots: new Map([[stateRootHex, parsedState]]),
					...(await getStateManagerOpts()),
				})
				await manager.ready()
				return manager
			}
			const genesisState = {
				...GENESIS_STATE,
				// add predeploys to genesis state
				...Object.fromEntries(
					(options.customPredeploys ?? []).map((predeploy) => [
						predeploy.contract.address,
						{
							nonce: 0n,
							deployedBytecode: predeploy.contract.deployedBytecode,
							// TODO this is definitely wrong but should work
							storageRoot: bytesToHex(KECCAK256_RLP),
							codeHash: keccak256(predeploy.contract.deployedBytecode),
						},
					]),
				),
			}
			const opts = await getStateManagerOpts()
			return createStateManager({
				...opts,
				currentStateRoot: stateRootHex,
				stateRoots: new Map([[bytesToHex(headBlock.header.stateRoot), genesisState]]),
				genesisState,
			})
		})

	const evmPromise = Promise.all([chainCommonPromise, stateManagerPromise, blockchainPromise]).then(
		([common, stateManager, blockchain]) => {
			return createEvm({
				common,
				stateManager,
				blockchain,
				allowUnlimitedContractSize: options.allowUnlimitedContractSize ?? false,
				customPrecompiles: options.customPrecompiles ?? [],
				profiler: options.profiler ?? false,
				loggingLevel,
			})
		},
	)

	const vmPromise = Promise.all([evmPromise, chainCommonPromise]).then(([evm, common]) => {
		const vm = createVm({
			stateManager: evm.stateManager,
			evm: evm,
			// TODO this inherited type is wrong thus we need to cast this
			blockchain: /** @type {import('@tevm/blockchain').Chain} */ (evm.blockchain),
			common,
		})
		return vm
	})

	const txPoolPromise = vmPromise.then((vm) => new TxPool({ vm }))
	const receiptManagerPromise = vmPromise.then((vm) => {
		logger.debug('initializing receipts manager...')
		return new ReceiptsManager(createMapDb({ cache: new Map() }), vm.blockchain)
	})
	/**
	 * @type {Map<import('viem').Hex, import('./Filter.js').Filter>}
	 */
	const filters = new Map()

	const createEventEmitter = () => {
		/** @type {Map<string | symbol, Array<Function>>} */
		const events = new Map()
		/**
		 * @type {import('./EIP1193EventEmitterTypes.js').EIP1193EventEmitter}
		 */
		const eventEmitter = {
			on(eventName, listener) {
				const listeners = events.get(eventName) || []
				listeners.push(listener)
				events.set(eventName, listeners)
			},
			removeListener(eventName, listener) {
				const listeners = events.get(eventName)
				if (listeners) {
					const index = listeners.findIndex((l) => l === listener)
					if (index !== -1) {
						listeners.splice(index, 1)
						if (listeners.length === 0) {
							events.delete(eventName)
						}
					}
				}
			},
			emit(eventName, ...args) {
				const listeners = events.get(eventName)
				if (listeners?.length) {
					listeners.forEach((listener) => listener(...args))
					return true // Event was successfully emitted
				}
				return false // No listeners for the event
			},
		}
		return eventEmitter
	}
	const eventEmitter = createEventEmitter()

	const readyPromise = (async () => {
		await blockchainPromise.then((b) => b.ready())
		await stateManagerPromise.then((b) => b.ready())
		await vmPromise.then((vm) => vm.ready())
		eventEmitter.emit('connect')
		return /** @type {true}*/ (true)
	})()

	/**
	 * @param {import('./TevmNode.js').TevmNode} baseClient
	 * @returns {import('./TevmNode.js').TevmNode['deepCopy']}
	 */
	const deepCopy = (baseClient) => async () => {
		/**
		 * @type {import('@tevm/utils').Address | undefined}
		 */
		let impersonatedAccount = undefined
		/**
		 * @param {import('@tevm/utils').Address | undefined} address
		 * returns {void}
		 */
		const setImpersonatedAccount = (address) => {
			impersonatedAccount = address
		}
		await readyPromise
		const oldVm = await vmPromise
		const vm = await oldVm.deepCopy()
		const oldReceiptsManager = await receiptManagerPromise
		const receiptsManager = oldReceiptsManager.deepCopy(vm.blockchain)
		const oldTxPool = await txPoolPromise
		const txPool = oldTxPool.deepCopy({ vm })
		/**
		 * @type {Map<import('viem').Hex, import('./Filter.js').Filter>}
		 */
		const newFilters = new Map(filters)
		// don't copy registered events because that would be confusing
		const eventEmitter = createEventEmitter()
		/**
		 * @type {import('./TevmNode.js').TevmNode}
		 */
		const copiedClient = {
			...eventEmitter,
			logger: baseClient.logger,
			getReceiptsManager: async () => Promise.resolve(receiptsManager),
			getTxPool: async () => Promise.resolve(txPool),
			getVm: async () => Promise.resolve(vm),
			miningConfig: baseClient.miningConfig,
			mode: baseClient.mode,
			...('forkTransport' in baseClient ? { forkTransport: baseClient.forkTransport } : {}),
			extend: (extension) => extend(baseClient)(extension),
			deepCopy: () => deepCopy(copiedClient)(),
			ready: () => Promise.resolve(true),
			getFilters: () => newFilters,
			getImpersonatedAccount() {
				return impersonatedAccount
			},
			setImpersonatedAccount,
			setFilter: (filter) => {
				newFilters.set(filter.id, filter)
			},
			removeFilter: (filterId) => {
				newFilters.delete(filterId)
			},
			status: 'READY',
		}
		return copiedClient
	}

	/**
	 * Create and return the baseClient
	 * It will be syncronously created but some functionality
	 * will be asyncronously blocked by initialization of vm and chainId
	 * @type {import('./TevmNode.js').TevmNode}
	 */
	const baseClient = {
		...eventEmitter,
		logger,
		getReceiptsManager: async () => {
			await readyPromise
			return receiptManagerPromise
		},
		getTxPool: async () => {
			await readyPromise
			return txPoolPromise
		},
		getVm: async () => {
			await readyPromise
			return vmPromise
		},
		miningConfig: options.miningConfig ?? { type: 'manual' },
		mode: options.fork?.transport ? 'fork' : 'normal',
		...(options.fork?.transport ? { forkTransport: options.fork.transport } : {}),
		extend: (extension) => extend(baseClient)(extension),
		ready: () => readyPromise,
		getImpersonatedAccount() {
			return impersonatedAccount
		},
		setImpersonatedAccount,
		getFilters: () => filters,
		setFilter: (filter) => {
			filters.set(filter.id, filter)
		},
		removeFilter: (filterId) => {
			filters.delete(filterId)
		},
		status: 'INITIALIZING',
		deepCopy: () => deepCopy(baseClient)(),
	}

	eventEmitter.on('connect', () => {
		if (baseClient.status !== 'INITIALIZING') {
			return
		}
		baseClient.status = 'READY'
	})

	return baseClient
}
