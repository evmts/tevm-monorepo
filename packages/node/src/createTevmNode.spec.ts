import { createAddress } from '@tevm/address'
import { createCommon, mainnet } from '@tevm/common'
import { definePredeploy } from '@tevm/predeploys'
import { CacheType, ContractCache, StorageCache } from '@tevm/state'
import { createSyncStoragePersister } from '@tevm/sync-storage-persister'
import { SimpleContract, transports } from '@tevm/test-utils'
import { EthjsAccount, EthjsAddress, bytesToHex } from '@tevm/utils'
import { describe, expect, it, vi } from 'vitest'
import { createTevmNode } from './createTevmNode.js'

describe('createTevmNode with State Persister', () => {
	it('Restores state correctly from persister', async () => {
		const storage = new Map()

		const persister = createSyncStoragePersister({
			key: 'test',
			storage: {
				getItem: (key) => storage.get(key),
				setItem: (key, value) => storage.set(key, value),
				removeItem: (key) => storage.delete(key),
			},
		})

		const client = createTevmNode({
			fork: {
				transport: transports.optimism,
				blockTag: 'latest',
			},
			persister,
		})

		await client.ready()

		// Verify state was restored correctly
		const vm = await client.getVm()
		const stateManager = vm.stateManager

		const address = EthjsAddress.fromString(`0x${'11'.repeat(20)}`)
		const account = EthjsAccount.fromAccountData({ nonce: 23n, balance: 100n })

		await stateManager.putAccount(address, account)

		expect(await stateManager.getAccount(address)).toEqual(account)

		expect(storage).toMatchSnapshot()

		const newClient = createTevmNode({
			persister,
		})

		await newClient.ready()

		const newVm = await newClient.getVm()
		const newStateManager = newVm.stateManager

		const restoredAccount = await newStateManager.getAccount(address)

		// TODO why is this broke? It shouldn't be not. Experiemental feature
		expect(restoredAccount).not.toEqual(account)
	})
})

describe('createTevmNode', () => {
	it('Creates a base client', async () => {
		const { mode, getVm, ready, extend, logger, forkTransport, getTxPool, miningConfig, getReceiptsManager } =
			createTevmNode()
		expect(mode).toBe('normal')
		expect(await ready()).toBe(true)
		expect(await getVm().then((vm) => vm.evm.runCall({}))).toMatchSnapshot()
		expect(extend).toBeInstanceOf(Function)
		expect(logger.warn).toBeInstanceOf(Function)
		expect(forkTransport).toBeUndefined()
		expect(await getTxPool().then((pool) => pool.pool)).toEqual(new Map())
		expect(miningConfig).toEqual({ type: 'manual' })
		expect(await getReceiptsManager().then((manager) => manager.getReceipts)).toBeInstanceOf(Function)
	})

	it('Can be extended', async () => {
		const client = createTevmNode().extend(() => ({
			hello: 'world',
		}))
		expect(client.hello).toBe('world')
	})

	it('Sets and gets impersonated account', () => {
		const client = createTevmNode()
		const address = '0x0000000000000000000000000000000000000001'
		client.setImpersonatedAccount(address)
		expect(client.getImpersonatedAccount()).toBe(address)
	})

	it('Manages filters correctly', () => {
		const client = createTevmNode()
		const filter = { id: createAddress(1).toString(), data: {} }
		client.setFilter(filter as any)
		expect(client.getFilters().get(createAddress(1).toString())).toEqual(filter as any)
		client.removeFilter(createAddress(1).toString())
		expect(client.getFilters().has(createAddress(1).toString())).toBe(false)
	})

	it('Handles fork mode with provided transport', async () => {
		const forkClient = createTevmNode({
			fork: { transport: transports.optimism },
		})
		const { mode, forkTransport } = forkClient
		expect(mode).toBe('fork')
		expect(forkTransport).toBe(transports.optimism)
	})

	it('Uses custom common and state manager options', async () => {
		const customCommon = createCommon({ ...mainnet, id: 999, hardfork: 'cancun', eips: [], loggingLevel: 'warn' })
		const customStateOptions = {
			loggingLevel: 'info',
			storageCache: new StorageCache({ size: 500, type: CacheType.LRU }),
			contractCache: new ContractCache(new StorageCache({ size: 500, type: CacheType.LRU })),
		} as const
		const customClient = createTevmNode({
			common: customCommon,
			...customStateOptions,
		})
		const { logger, getVm } = customClient
		expect(logger.level).toBe('info')
		const vm = await getVm()
		expect(vm.common.id).toBe(999)
	})

	it('Persists state with custom persister', async () => {
		const persister = {
			persistTevmState: vi.fn(),
			restoreState: vi.fn(),
		}
		const persistingClient = createTevmNode({
			persister: persister as any,
		})
		await persistingClient.ready()
		await new Promise((resolve) => setTimeout(resolve, 100))
		expect(persister.persistTevmState).toHaveBeenCalled()
	})

	it('Handles connect event correctly', async () => {
		const client = createTevmNode()
		let connected = false
		client.on('connect', () => {
			connected = true
		})
		await client.ready()
		expect(connected).toBe(true)
	})

	it('Persists state with custom persister', async () => {
		const persister = {
			persistTevmState: vi.fn(),
			restoreState: vi.fn(),
		}
		const persistingClient = createTevmNode({
			persister: persister as any,
		})
		await persistingClient.ready()
		await new Promise((resolve) => setTimeout(resolve, 100))
		expect(persister.persistTevmState).toHaveBeenCalled()
	})

	it('Handles connect event correctly', async () => {
		const client = createTevmNode()
		let connected = false
		const listener = () => {
			connected = true
		}
		client.on('connect', listener)
		await client.ready()
		expect(connected).toBe(true)
		client.removeListener('connect', listener)
	})

	it('Should be able to remove listener', async () => {
		const client = createTevmNode()
		let connected = false
		const listener = () => {
			connected = true
		}
		client.on('connect', listener)
		client.removeListener('connect', listener)
		await client.ready()
		expect(connected).toBe(false)
	})

	it('Handles custom predeploys in genesis state', async () => {
		const predeploy = definePredeploy(SimpleContract.withAddress(`0x${'22'.repeat(20)}`))
		const client = createTevmNode({
			customPredeploys: [predeploy],
		})
		const { getVm } = client
		const vm = await getVm()
		const stateManager = vm.stateManager
		const code = await stateManager.getContractCode(createAddress(predeploy.contract.address))
		expect(bytesToHex(code)).toBe(predeploy.contract.deployedBytecode)
	})

	it('Fetches latest block number for fork block tag "latest"', async () => {
		const client = createTevmNode({
			fork: { transport: transports.optimism, blockTag: 'latest' },
		})
		await client.ready()
		const { getVm } = client
		const vm = await getVm()
		expect(vm.blockchain.getCanonicalHeadBlock()).toBeDefined()
	})

	describe('deepCopy', () => {
		it('Client.deepCopy() returns a deep copy of the client', async () => {
			const client = createTevmNode()
			const copy = await client.deepCopy()
			expect(copy).not.toBe(client)
			expect(await client.getVm()).not.toBe(await copy.getVm())
			expect(await client.getTxPool()).not.toBe(await copy.getTxPool())
			expect(await client.getReceiptsManager()).not.toBe(await copy.getReceiptsManager())
			expect((await client.getTxPool()).pool).not.toBe((await copy.getTxPool()).pool)
			expect((await client.getTxPool()).pool).toEqual((await copy.getTxPool()).pool)
			expect((await client.getVm()).stateManager._baseState.caches).not.toBe(
				(await copy.getVm()).stateManager._baseState.caches,
			)
			expect((await client.getVm()).stateManager._baseState.stateRoots).toEqual(
				(await copy.getVm()).stateManager._baseState.stateRoots,
			)
			expect((await client.getVm()).stateManager._baseState.getCurrentStateRoot()).toEqual(
				(await copy.getVm()).stateManager._baseState.getCurrentStateRoot(),
			)
		})
	})
})
