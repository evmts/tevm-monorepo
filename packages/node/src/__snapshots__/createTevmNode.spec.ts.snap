// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createTevmNode > Creates a base client 1`] = `
{
  "createdAddress": Address {
    "bytes": Uint8Array [
      189,
      119,
      4,
      22,
      163,
      52,
      95,
      145,
      228,
      179,
      69,
      118,
      203,
      128,
      74,
      87,
      111,
      164,
      142,
      177,
    ],
  },
  "execResult": {
    "exceptionError": undefined,
    "executionGasUsed": 0n,
    "gasRefund": 0n,
    "returnValue": Uint8Array [],
  },
}
`;

exports[`createTevmNode with State Persister > Restores state correctly from persister 1`] = `Map {}`;
